x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/BaseSearchableSnapshotIndexInput.java

```diff
@@ -300,9 +300,12 @@ protected final boolean assertCurrentThreadMayAccessBlobStore() {
             // Cache prewarming also runs on a dedicated thread pool.
             || threadName.contains('[' + SearchableSnapshots.CACHE_PREWARMING_THREAD_POOL_NAME + ']')
 
-            // Unit tests access the blob store on the main test thread; simplest just to permit this rather than have them override this
+            // Unit tests access the blob store on the main test thread, or via an asynchronous
+            // checkindex call;
+            // simplest just to permit this rather than have them override this
             // method somehow.
             || threadName.startsWith("TEST-")
+            || threadName.startsWith("async-check-index")
```

Comment:
We should not read directly from async-check-index unless the current IndexInput is an instance of DirectBlobContainerIndexInput or we're reading bytes directly using readDirectlyIfAlreadyClosed() because we failed to read bytes from caches. 

I wonder if we could pass a `direct` parameter to `assertCurrentThreadMayAccessBlobStore()` and be more permissive on thread names when `direct` is `true`.

Reply:
It looks like in this case it was executed from `MetadataCachingIndexInput`:

```
6 of 7: name=_0 maxDoc=358
    version=8.11.1
    id=enniyhmt78i2tmupc6p6pxjmg
    codec=Asserting(Lucene87)
    compound=true
    numFiles=6
    size (MB)=0.019
    diagnostics = {java.vendor=Eclipse Adoptium, os=Linux, java.version=17.0.1, java.vm.version=17.0.1+12, lucene.version=8.11.1, os.arch=amd64, java.runtime.version=17.0.1+12, source=flush, os.version=5.4.0-91-generic, timestamp=1641308531353}
    no deletions
    test: open reader.........OK [took 0.004 sec]
    test: check integrity.....FAILED
    WARNING: exorciseIndex() would remove reference to this segment; full exception:
java.lang.AssertionError: current thread [Thread[async-check-index-21-thread-2,5,TGRP-FrozenSearchableSnapshotsIntegTests]] may not read [name: __V82xs9_mTkOhc2Eum7lLCA, numberOfParts: 21, partSize: 834b, partBytes: 834, metadata: name [_0.cfs], length [16811], checksum [1xujv37], writtenBy [8.11.1]]
	at org.elasticsearch.xpack.searchablesnapshots.store.input.BaseSearchableSnapshotIndexInput.assertCurrentThreadMayAccessBlobStore(BaseSearchableSnapshotIndexInput.java:306)
	at org.elasticsearch.xpack.searchablesnapshots.store.input.BaseSearchableSnapshotIndexInput.openInputStreamFromBlobStore(BaseSearchableSnapshotIndexInput.java:176)
	at org.elasticsearch.xpack.searchablesnapshots.store.input.MetadataCachingIndexInput.readDirectlyIfAlreadyClosed(MetadataCachingIndexInput.java:383)
	at org.elasticsearch.xpack.searchablesnapshots.store.input.MetadataCachingIndexInput.doReadInternal(MetadataCachingIndexInput.java:111)
	at org.elasticsearch.xpack.searchablesnapshots.store.input.BaseSearchableSnapshotIndexInput.readInternal(BaseSearchableSnapshotIndexInput.java:112)
	at org.apache.lucene.store.BufferedIndexInput.readBytes(BufferedIndexInput.java:152)
	at org.apache.lucene.store.BufferedIndexInput.readBytes(BufferedIndexInput.java:111)
	at org.apache.lucene.store.BufferedChecksumIndexInput.readBytes(BufferedChecksumIndexInput.java:49)
	at org.apache.lucene.store.DataInput.readBytes(DataInput.java:88)
	at org.apache.lucene.store.DataInput.skipBytes(DataInput.java:338)
	at org.apache.lucene.store.ChecksumIndexInput.seek(ChecksumIndexInput.java:52)
	at org.apache.lucene.codecs.CodecUtil.checksumEntireFile(CodecUtil.java:546)
	at org.apache.lucene.codecs.blocktree.BlockTreeTermsReader.checkIntegrity(BlockTreeTermsReader.java:349)
	at org.apache.lucene.codecs.perfield.PerFieldPostingsFormat$FieldsReader.checkIntegrity(PerFieldPostingsFormat.java:371)
	at org.apache.lucene.index.CodecReader.checkIntegrity(CodecReader.java:285)
	at org.apache.lucene.index.SegmentReader.checkIntegrity(SegmentReader.java:372)
	at org.apache.lucene.index.CheckIndex.testSegment(CheckIndex.java:921)
	at org.apache.lucene.index.CheckIndex.lambda$checkIndex$1(CheckIndex.java:765)
	at org.apache.lucene.index.CheckIndex.lambda$callableToSupplier$2(CheckIndex.java:850)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

As discussed with @romseygeek we should instead configure index checker to use single/caller thread to perform index check for now to preserve current behavior (before lucene 8.11.x). I will open another pr shortly.

Reply:
> ```
> 
> ```
> 
> 
>     
>   
> 
> As discussed with @romseygeek we should instead configure index checker to use single/caller thread to perform index check for now to preserve current behavior (before lucene 8.11.x).

+1